TAP-2.md

```diff
@@ -37,123 +37,84 @@ Provide all on-chain liquidity in a familiar limit orderbook interface to read (
-  maxSplits?: number
-  includedProtocols?: string[]
-  pricePoints?: number
-  gasMode: GasMode
-  gasToken?: string
-  maxGasPriceGwei?: string
-}
-
-type FinetuneOptions = OrderbookOptions & {
-  tolerance: string
-  maxIterations?: number
+pub struct ProtocolStreamManager {
+    tycho_url: String,
+    auth_key: Option<String>,
+    tvl_threshold: f64,
 }
+```
+### 2. Router Interface
+An interface for routing solutions. Implement a simple solver, and expect to later integrate a more sophisticated external solver.
+```rust
+/// Generic router interface for finding and executing trades
+pub trait Router {
+    fn get_amount_out(
+        &self,
+        amount_in: U256,
+        token_in: &Address,
+        token_out: &Address,
+    ) -> Result<GetAmountOutResult, Error>;
 
-// Operation Results
-type SizeQuote = {
-  effectivePrice: string
-  route: Route
-  gasCost: Level['gasCost']
-  priceImpact: string
+    fn encode_solution(
+        &self,
+        result: GetAmountOutResult,
+        sender: Address,
+        receiver: Address,
+    ) -> Result<RoutingSolution, Error>;
 }
-
-type DepthAtPrice = {
-  totalSize: string
-  averagePrice: string
-  route: Route
-  gasCost: Level['gasCost']
+```
+### 3. Price Level Calculator 
+Calculates discrete price levels for the orderbook using router simulations.
+```rust
+pub struct PriceLevel {
+    pub price: U256,
+    pub size: U256,
+    pub solution: RoutingSolution,
 }
 
-// Event System
-type OrderbookEvent = {
-  type: 'price_level_changed'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  side: Side
-  price: string
-  newSize: string
-} | {
-  type: 'book_reset'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
+pub struct PriceLevelCalculator {
+    router: Box<dyn Router>,
+    price_points: u32,
 }
-
-// Core Operations
-type Operations = {
-  getOrderbook: (pair: TokenPair, options?: OrderbookOptions) => Promise<OrderbookState>
-  
-  getQuoteForSize: (
-    pair: TokenPair,
-    size: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<SizeQuote>
-  
-  getSizeAtLimit: (
-    pair: TokenPair,
-    limitPrice: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<DepthAtPrice>
-  
-  findOptimalSize: (
-    pair: TokenPair,
-    targetPrice: string,
-    options?: FinetuneOptions
-  ) => Promise<SizeQuote>
-  
-  subscribeToEvents: (
+```
+### 4. Orderbook Manager
+Maintains orderbook state and handles updates as new protocol states arrive.
+```rust
+pub struct OrderbookState {
     pair: TokenPair,
-    onEvent: (event: OrderbookEvent) => void
-  ) => { unsubscribe: () => void }
-  
-  getSupportedTokens: () => Promise<string[]>
-  getSupportedPairs: () => Promise<TokenPair[]>
+    timestamp: u64,
+    block_number: u64,
+    bids: Vec<PriceLevel>,
+    asks: Vec<PriceLevel>,
+}
+pub struct OrderbookManager {
+    calculator: PriceLevelCalculator,
+    current_books: HashMap<TokenPair, OrderbookState>,
```

Comment:
Ok I think this should be somehow public but I am not sure if the state should be public, instead I think you should specify a struct with only methods. The methods then return whatever you would like to have, e.g. orderbook state for a pair...